bugsnag-android-core/src/main/java/com/bugsnag/android/DocumentPath.java

```diff
@@ -0,0 +1,385 @@
+package com.bugsnag.android;
+
+import android.icu.lang.UCharacterEnums;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * DocumentPath records a path into a hierarchical acyclic document, and can be used to modify
+ * documents by navigating down the DAG and then applying a modification at that level.
+ *
+ * Paths are explained as follows in PLAT-6353:
+ *
+ * Paths follow a dotted notation similar to internet hostnames, such that each path component
+ * is separated by a dot '.' If a literal dot is required, the path supports an escaping
+ * mechanism using the backslash character:
+ *
+ *     \. = literal dot
+ *     \\ = literal backslash
+ *
+ * Path components are either integers or non-integers. Integers refer to the previous path
+ * component as a list, providing the index to look up. Non-integers refer to the previous path
+ * component as a map, providing the key to look up (for example, in a.b.3, a is treated as a
+ * map and b is treated as a list, so the path refers to document root with map key a, which is
```

Comment:
It would be incredibly helpful to define 'a', 'b', etc as constants so that it's easier to understand what's going on in the tests.

Reply:
I'm not sure what you mean... 'a', 'b', etc are just examples to help explain how paths work. They don't actually exist.